crypto/blake256/internal/_asm/gen_compress_asm_amd64.go

```diff
@@ -0,0 +1,1109 @@
+			build.VMOVDQA(rows[1], hr[1])                    //  hr1  = h4 h5 h6 h7
+		},
+		convertMsgFn: func() {
+			// The message is stored in the XMM registers as:
+			// msg0 = |m0  m1  m2  m3|
+			// msg1 = |m4  m5  m6  m7|
+			// msg2 = |m8  m9  ma  mb|
+			// msg3 = |mc  md  me  mf|
+			build.VMOVDQU(globals.leToBe4x32, shuf)
+			for i := 0; i < 4; i++ {
+				build.VMOVDQU(memOffset(msgPtr, i*16), msg[i])
+				build.VPSHUFB(shuf, msg[i], msg[i])
+			}
+		},
+		columnStepFn: func(round uint8, isDiagStep bool) {
+			// Determine the memory offsets for the permuted constants for this
+			// round and column/diagonal step.
+			baseConstsOffset := int(round%10) * 64
+			if isDiagStep {
+				baseConstsOffset += 32
+			}
+			cxOffset := globals.permutedConsts.Offset(baseConstsOffset)
+			cyOffset := globals.permutedConsts.Offset(baseConstsOffset + 16)
+
+			loadMsg4x32(round, isDiagStep, true, imr[0])  // imr0 = mx (for G[i]..G[i+3])
+			build.VMOVDQU(cxOffset, imr[1])               // imr1 = cx (for G[i]..G[i+3])
+			build.VPXOR(imr[0], imr[1], imr[1])           // imr1 = mx^cx
+			build.VPADDD(imr[1], rows[0], rows[0])        // row0 += imr1 (mx^cx)
+			loadMsg4x32(round, isDiagStep, false, imr[0]) // imr0 = my (for G[i]..G[i+3])
+			build.VMOVDQU(cyOffset, imr[1])               // imr1 = cy (for G[i]..G[i+3])
+			build.VPXOR(imr[0], imr[1], imr[1])           // imr1 = my^cy
+			build.VPADDD(rows[1], rows[0], rows[0])       // row0 += row1
+			build.VPXOR(rows[0], rows[3], rows[3])        // row3 ^= row0
+			rotateRight4x32AVX(16, rows[3])               // row3 >>>= 16
+			build.VPADDD(rows[3], rows[2], rows[2])       // row2 += row3
+			build.VPXOR(rows[2], rows[1], rows[1])        // row1 ^= row2
+			rotateRight4x32AVX(12, rows[1])               // row1 >>>= 12
+			build.VPADDD(imr[1], rows[0], rows[0])        // row1 += imr1 (my^cy)
+			build.VPADDD(rows[1], rows[0], rows[0])       // row0 += row1
+			build.VPXOR(rows[0], rows[3], rows[3])        // row3 ^= row0
+			rotateRight4x32AVX(8, rows[3])                // row3 >>>= 8
+			build.VPADDD(rows[3], rows[2], rows[2])       // row2 += row3
+			build.VPXOR(rows[2], rows[1], rows[1])        // row1 ^= row2
+			rotateRight4x32AVX(7, rows[1])                // row1 >>>= 7
+		},
+		diagonalizeFn:   diagonalizeAVX,
+		undiagonalizeFn: undiagonalizeAVX,
+		finalizeFn: func() {
+			build.VPXOR(hr[0], rows[0], rows[0])   // row0 ^= h0 h1 h2 h3
+			build.VPXOR(sr, rows[0], rows[0])      // row0 ^= s0 s1 s2 s3
+			build.VPXOR(rows[2], rows[0], rows[0]) // row0 ^= v8 v9 va vb
+			build.VPXOR(hr[1], rows[1], rows[1])   // row1 ^= h4 h5 h6 h7
+			build.VPXOR(sr, rows[1], rows[1])      // row1 ^= s0 s1 s2 s3
+			build.VPXOR(rows[3], rows[1], rows[1]) // row1 ^= vc vd ve vf
+		},
+		outputChainValueFn: func() {
+			build.VMOVDQU(rows[0], mustAddr(h.Index(0))) // h0 h1 h2 h3 = row0
+			build.VMOVDQU(rows[1], mustAddr(h.Index(4))) // h4 h5 h6 h7 = row1
+		},
+	}
+	compressor.generate(rows)
```

Comment:
A call to `build.VZEROUPPER()` is usually recommended before exiting AVX code, even though there's little chance of mixing SSE and AVX code here, and this code is not using the upper bits of the YMM registers anyway.

Reply:
`VZEROUPPER` is called from the AVX code.

Reply:
Oh, that was for AVX2 where I had it.

I removed it from the AVX because it doesn't use any of the upper bits of the YMM registers, so they're all guaranteed to be zero and thus I didn't think it necessary.